### DISCUSSION TITLE
Roadmap or expectations for Fabric support and V3.

### DISCUSSION THREAD

**Discussion by AndreiCalazans** [software-mansion/react-native-reanimated#3529] [General]
Hi, thank you for your work on Reanimated. I couldn't find any roadmaps or set of expectations for when V3 will have a stable release with support for Fabric. What are the expectations around this? Is there a place we can track? 

Thank you again.

**Comment by psycheangel** [comment]
Hi, also there is no follow-up on how to build from the source for reanimated v3.

I have tested v3 using aar RN 70 (with RN init cra) seems to be working in debug mode, but for the release version of apk still have crash issue

**Reply by tomekzaw** [reply]
> Hi, also there is no follow-up on how to build from the source for reanimated v3.

Do you mean compiling Reanimated C++ codebase when building the app? Reanimated v3 is only built from sources (there are no AARs in the npm package) and thus it requires Android NDK to be installed and properly set up.

> but for the release version of apk still have crash issue

Can you please submit an issue with the full stack trace of this crash you are experiencing?

**Reply by psycheangel** [reply]
@tomekzaw its probably because I`m not building reanimated v3 from the source. 
is the command for the building would be /gradlew installArchives ?

I need to clarify, the AAR is used because there is other package library using it (rn-vision-camera)

**Reply by tomekzaw** [reply]
Hey, I'm not sure whether react-native-vision-camera supports Reanimated 3. cc @mrousavy

**Reply by psycheangel** [reply]
Yes, it's currently not supporting fabric, I use the AAR to skip compile failure (not for production usage). 
the crash is not an issue with reanimated v3 package, I already found out why it crashed. 
thanks @tomekzaw

**Reply by mrousavy** [reply]
Hey! Yea VisionCamera does currently not work with the new React Native Architecture. I didn't have the time to look into this yet, so I would appreciate any help there. It's on my todo list tho, the new arch will bring tons of improvements and new possibilities to the next version of VisionCamera - I'll also make sure to make it compatible with REA v3.

**Comment by tomekzaw** [comment]
> I couldn't find any roadmaps or set of expectations for when V3 will have a stable release with support for Fabric. What are the expectations around this?

The plan is that we will release 3.0.0 stable once we implement Layout Animations for Fabric. Apart from this, we are still working on profiling and optimizing algorithms related to animating layout props.

Currently, Reanimated 3.0.0-rc.x supports Fabric on RN 0.69+. However, due to a number of breaking changes in recent releases of React Native, we are considering dropping support for RN <=0.70 once RN 0.71 rolls out with prefabs which may greatly simplify the build scripts for Android. Right now we have a number of giant chunks of code in for downloading native dependencies (boost, glog, folly, double-conversion), unpacking *.so files from *.aar etc. that we really wish to get rid of. Also, when releasing a new version of Reanimated, we need to test the package against a wide range of RN versions.

> Is there a place we can track?

I think that this discussion might be such place 😄

**Comment by capezzbr** [comment]
> However, due to a number of breaking changes in recent releases of React Native, we are considering dropping support for RN <=0.70 once RN 0.71 rolls out with prefabs which may greatly simplify the build scripts for Android.

Does that mean that there is a world where Reanimated 3.0.0 stable will be released with support for RN >= 0.69, and then follow up versions will only be compatible with RN > 0.70 (once RN 0.71 is released), or that Reanimated 3.0.0 stable will only happen for RN 0.71?

I am asking since RN 0.70 was just released and I was hoping that Reanimated 3.0.0 stable will be compatible with that 🙇

**Reply by tomekzaw** [reply]
**TL;DR** Reanimated 3.0.0-rc.2 supports RN 0.62+ on Paper and RN 0.69+ on Fabric. For 3.0.0-stable, we are considering dropping support for RN 0.62-0.68 on Paper so v3 would support RN 0.69+ (Paper & Fabric). Please please migrate your apps to newer versions of RN because 0.62 is like 2,5 years old (it was released on 26 Mar 2020). 

In [Announcing Reanimated 3](https://blog.swmansion.com/announcing-reanimated-3-16167428c5f7) blog post we laid out a complete list of changes between v2 and v3:
* ✂️ Drop Reanimated 1 API  
* 🏗️ No longer distributed as AAR (build from source only)  
* ⭐ Fabric support (RN 0.69+)

Basically, we promised that everyone who uses Reanimated v2 and does not use v1 API can safely upgrade to v3. In particular, this means we still need to support RN 0.62+ on Paper.

Since the New Architecture is still experimental, we decided to officially support Fabric only the latest stable release of RN, i.e. RN 0.69 for 3.0.0-rc.0 and RN 0.70 for 3.0.0-rc.2. Don't get me wrong, rc.2 still works fine with RN 0.69, but we make no promises here 😄 

When it comes to Paper (the old architecture), the thing is that there are 10 major versions of React Native between 0.62 and 0.71 and each brings its own changes that we need to take into account. Also, there are major differences related to how JSI, Hermes, Folly and other RN internals are distributed which make our `build.gradle` and `CMakeLists.txt` more difficult to maintain. Finally, each new release has to be tested against multiple possible versions of RN.

Right now we are considering dropping support for RN 0.62-0.68 so Reanimated v3 supports RN 0.69+ both on Paper and Fabric. Not only this would greatly simplify our build scripts for Android but also we wouldn't have to worry that we introduce some feature that fails to compile on let's say RN 0.63. 

We also have this cool idea of dropping support for RN <=0.70 on Paper and thus support only RN 0.71+ (Paper & Fabric) once the native dependencies are published as prefabs. It would make sense to introduce such a big change with new major release (i.e. Reanimated 3). However, after some consideration, it might be better to wait for some time before we make such decision and ask the community if it's ready for such a giant leap.

**Reply by capezzbr** [reply]
Thanks for clarifying. I believe that Reanimated v3 stable supporting RN 0.69+ (Paper & Fabric) is a good compromise, since RN 0.69 has been out for a bit, so it'll be easier for folks to be on that version. 

> We also have this cool idea of dropping support for RN <=0.70 on Paper and thus support only RN 0.71+ (Paper & Fabric) once the native dependencies are published as prefabs

I believe this option will make it harder for apps to start exploring/testing with the new architecture, as it will now require apps to be on RN 0.71 to have a stable version of Reanimated that supports Fabric. My vote goes to having Reanimated v3 stable supporting RN 0.69+, since it is a good compromise (e.g. it drops support for RN 0.68, but doesn't require you to be on the next version of RN).

**Reply by mrousavy** [reply]
> We also have this cool idea of dropping support for RN <=0.70 on Paper and thus support only RN 0.71+ (Paper & Fabric) once the native dependencies are published as prefabs

I think this is a good idea. As library authors it's really hard to maintain support for every RN version, and I think a jump from RN 0.68 to RN 0.71 isn't too unrealistic for most apps. I am not using the new arch in any of my apps yet, and I doubt a lot of people are.

Using prefabs makes everything way easier and ideally build faster as well, I'll definitely go that path with my libraries 😄

**Reply by capezzbr** [reply]
> I am not using the new arch in any of my apps yet, and I doubt a lot of people are.

That is probably true, but having to be on RN 0.71 in order to have a stable version of Reanimated V3, will make this even worse 😅

**Reply by mrousavy** [reply]
Well yea, but if you take a look at the CMakeLists.txt / build.gradle files of REA or VisionCamera or MMKV you'll probably understand that this is a fair compromise - it's just really hard for library authors to maintain support for everything, and waiting to enable the new arch until you're on RN 0.71 is also a fair assumption imo.

**Reply by fergusmeiklejohn** [reply]
I think you should drop support for versions of RN which significantly affect your ability to maintain and improve Reanimatedv3. My reasoning is this: if a project has to use or support an old version of RN then that project can use ReanimatedV2, which already works brilliantly.

**Reply by capezzbr** [reply]
Given that React Native 0.71 will be released soon, I am in favor of what said above 😃  Especially if that means that Reanimated V3 will be able to receive more support and iterate faster!
